### ISSUE TITLE
stdin inside container intermittently not being closed when attached stdin is closed

### ISSUE THREAD

**Issue by joeshaw** [docker/docker#29421]
**Description**

I am running a container that is reading from stdin.  When I close stdin from `docker run` it doesn't always results in stdin in the container being closed.

**Steps to reproduce the issue:**

The image used is based on `ubuntu:12.04` with some packages installed, including strace.

`docker run -i -a stdin -a stderr myimage strace cat -`

You'll see some strace output, and it will be waiting in a `read()` system call for input.  You can type stuff in there and see it output, but that's not necessary.

Hit ctrl-D to close stdin.  Sometimes the `strace cat -` inside the container sees that stdin has been closed and the container exits, but other times it's left running with stdin still open.  The Docker daemon notices that the attached stdin has been closed and it becomes impossible to send additional input to the container.  (However, ctrl-C still works since that's a different API.)

Also of note:  it seems that running `echo "foo" | docker run -i -a stdin -a stderr myimage strace cat -` always succeeds.

**Describe the results you received:**

Running the daemon with `-D`, in all cases (both success and failure) I see this in the Docker logs when I close stdin:

```
time="2016-12-14T19:44:34.791584553Z" level=debug msg="Calling POST /v1.24/containers/f80d2ef618828107c017121eece75c0758ce3fe6d323941661cd62ade05b1f12/attach?stderr=1&stdin=1&stream=1"
time="2016-12-14T19:44:34.791767074Z" level=debug msg="attach: stdin: begin"
time="2016-12-14T19:44:34.791847545Z" level=debug msg="attach: stderr: begin"
time="2016-12-14T19:44:34.792256224Z" level=debug msg="Calling POST /v1.24/containers/f80d2ef618828107c017121eece75c0758ce3fe6d323941661cd62ade05b1f12/start"
...
time="2016-12-14T19:45:03.428100292Z" level=debug msg="Closing buffered stdin pipe"
time="2016-12-14T19:45:03.428277792Z" level=debug msg="attach: stdin: end"
```

On a failure, this is the strace output:

```
brk(0)                                  = 0x17a3000
brk(0x17c4000)                          = 0x17c4000
fstat(1, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
fstat(0, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
fadvise64(0, 0, 0, POSIX_FADV_SEQUENTIAL) = -1 ESPIPE (Illegal seek)
read(0,
```
(hanging there)

On a success, this is the strace output:

```
brk(0)                                  = 0x1564000
brk(0x1585000)                          = 0x1585000
fstat(1, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
fstat(0, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
fadvise64(0, 0, 0, POSIX_FADV_SEQUENTIAL) = -1 ESPIPE (Illegal seek)
read(0, "", 32768)                      = 0
close(0)                                = 0
close(1)                                = 0
close(2)                                = 0
exit_group(0)                           = ?
```

In other words, the `read()` syscall completes by reading 0 bytes.

**Describe the results you expected:**

I expect the successful exit in all cases.

**Additional information you deem important (e.g. issue happens only occasionally):**

Happens roughly 50% of the time, except as I said the case where you pipe echo output works 100% of the time.  I have no idea why.

**Output of `docker version`:**

```
Client:
 Version:      1.12.4
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   1564f02
 Built:        Mon Dec 12 23:42:14 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.4
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   1564f02
 Built:        Mon Dec 12 23:42:14 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 54
 Running: 0
 Paused: 0
 Stopped: 54
Images: 15
Server Version: 1.12.4
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 151
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor
Kernel Version: 3.13.0-96-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.798 GiB
Name: dev
ID: JIMY:7XFY:IKHS:JZKL:RK6A:IJHE:SR46:MM3T:QEEP:HX4C:HHJY:UQW5
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 14
 Goroutines: 23
 System Time: 2016-12-14T19:53:35.869066869Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

This is run inside a vagrant VM on VMware Fusion on a Mac.

**Comment by joeshaw** [comment]
Possibly related to #29095, cc @tonistiigi 

I will try with 1.12.3.

**Comment by joeshaw** [comment]
Seems to work consistently with 1.12.3, so it looks like something from #29095 is probably the culprit.

**Comment by joeshaw** [comment]
Also related: https://github.com/docker/docker/pull/27405.  There's a comment in there about 3.13 kernels, which I am running on.

**Comment by tonistiigi** [comment]
This is fixed by https://github.com/docker/docker/pull/28682 . Should have been included in #29095

**Comment by thaJeztah** [comment]
Thanks for reporting @joeshaw, and thanks for doing the legwork on finding possible causes 👍

**Comment by justincormack** [comment]
This is fixed in 1.12.5 which should be released soon.

**Comment by thaJeztah** [comment]
docker 1.12.5 has been released, which contains https://github.com/docker/docker/pull/29424 to resolve this issue; https://github.com/docker/docker/releases/tag/v1.12.5, so I'm closing this
